File: zig/examples/cross_compat.zig
const std = @import("std");
const ason = @import("ason");
const print = std.debug.print;

const Detail = struct {
    ID: i64,
    Name: []const u8,
    Age: i32,
    Gender: bool,
};

const User = struct {
    details: []const Detail,
};

const Person = struct {
    ID: i64,
    Name: []const u8,
    Age: i32,
};

const Human = struct {
    details: []const Person,
};

pub fn main() !void {
    var gpa_impl: std.heap.GeneralPurposeAllocator(.{}) = .{};
    defer _ = gpa_impl.deinit();
    const gpa = gpa_impl.allocator();

    const users = [_]User{
        .{
            .details = &[_]Detail{
                .{ .ID = 1, .Name = "Alice", .Age = 30, .Gender = true },
                .{ .ID = 2, .Name = "Bob", .Age = 25, .Gender = false },
            },
        },
    };

    // Encode
    const ason_str = try ason.encode([]const User, &users, gpa);
    defer gpa.free(ason_str);
    print("Encoded ASON:\n{s}\n", .{ason_str});

    // Decode into Human
    const decoded = try ason.decode([]Human, ason_str, gpa);
    defer {
        for (decoded) |h| {
            for (h.details) |p| gpa.free(p.Name);
            gpa.free(h.details);
        }
        gpa.free(decoded);
    }

    print("\nDecoded into Human list:\n", .{});
    for (decoded) |h| {
        print("Human{{details=[", .{});
        for (h.details, 0..) |p, i| {
            if (i > 0) print(", ", .{});
            print("Person{{ID={d}, Name=\"{s}\", Age={d}}}", .{ p.ID, p.Name, p.Age });
        }
        print("]}}\n", .{});
    }
}

File: c/examples/cross_compat.c
#include "ason.h"
#include <assert.h>
#include <stdio.h>
#include <string.h>

// ----------------------------------------------------------------------------
// Generation structures
// ----------------------------------------------------------------------------
typedef struct {
  int64_t id;
  ason_string_t name;
  int32_t age;
  bool gender;
} Detail;

ASON_FIELDS(Detail, 4, ASON_FIELD(Detail, id, "ID", i64),
            ASON_FIELD(Detail, name, "Name", str),
            ASON_FIELD(Detail, age, "Age", i32),
            ASON_FIELD(Detail, gender, "Gender", bool))

ASON_VEC_STRUCT_DEFINE(Detail)

typedef struct {
  ason_vec_Detail details;
} User;

ASON_FIELDS(User, 1, ASON_FIELD_VEC_STRUCT(User, details, "details", Detail))

// ----------------------------------------------------------------------------
// Consumption structures
// ----------------------------------------------------------------------------
typedef struct {
  int64_t id;
  ason_string_t name;
  int32_t age;
} Person;

ASON_FIELDS(Person, 3, ASON_FIELD(Person, id, "ID", i64),
            ASON_FIELD(Person, name, "Name", str),
            ASON_FIELD(Person, age, "Age", i32))

ASON_VEC_STRUCT_DEFINE(Person)

typedef struct {
  ason_vec_Person details;
} Human;

ASON_FIELDS(Human, 1, ASON_FIELD_VEC_STRUCT(Human, details, "details", Person))

// ----------------------------------------------------------------------------
// Main
// ----------------------------------------------------------------------------
int main(void) {
  // 1. Setup User data
  User u = {0};
  u.details = ason_vec_Detail_new();

  Detail d1 = {1, ason_string_from("Alice"), 30, true};
  Detail d2 = {2, ason_string_from("Bob"), 25, false};

  ason_vec_Detail_push(&u.details, d1);
  ason_vec_Detail_push(&u.details, d2);

  User users[1] = {u};

  // 2. Encode
  ason_buf_t buf = ason_encode_vec_User(users, 1);
  printf("Encoded ASON:\n%.*s\n", (int)buf.len, buf.data);

  // 3. Decode into Human list
  Human *humans = NULL;
  size_t count = 0;
  ason_err_t err = ason_decode_vec_Human(buf.data, buf.len, &humans, &count);
  assert(err == ASON_OK);

  printf("\nDecoded into Human list:\n");
  for (size_t i = 0; i < count; i++) {
    printf("Human{details=[");
    for (size_t j = 0; j < humans[i].details.len; j++) {
      if (j > 0)
        printf(", ");
      printf("Person{ID=%lld, Name=\"%s\", Age=%d}",
             (long long)humans[i].details.data[j].id,
             humans[i].details.data[j].name.data,
             humans[i].details.data[j].age);
    }
    printf("]}\n");
  }

  // Cleanup
  ason_buf_free(&buf);
  for (size_t i = 0; i < count; i++) {
    for (size_t j = 0; j < humans[i].details.len; j++) {
      ason_string_free(&humans[i].details.data[j].name);
    }
    ason_vec_Person_free(&humans[i].details);
  }
  free(humans);

  ason_string_free(&d1.name);
  ason_string_free(&d2.name);
  ason_vec_Detail_free(&u.details);

  return 0;
}
